apps/language_server/lib/language_server/providers/rename.ex

```diff
@@ -0,0 +1,161 @@
+
+        definition_references =
+          case definition do
+            %{file: nil, type: :function} ->
+              parse_definition_source_code(source_file.text)
+              |> get_all_fn_header_positions(char_ident)
+              |> positions_to_references(start_uri, length_old)
+
+            %{file: separate_file_path, type: :function} ->
+              parse_definition_source_code(definition)
+              |> get_all_fn_header_positions(char_ident)
+              |> positions_to_references(SourceFile.Path.to_uri(separate_file_path), length_old)
+
+            _ ->
+              positions_to_references(
+                [{definition.line, definition.column}],
+                start_uri,
+                length_old
+              )
+          end
+
+        Enum.uniq(definition_references ++ repack_references(references, start_uri))
+      else
+        _ ->
+          []
+      end
+
+    changes =
+      edits
+      |> Enum.group_by(& &1.uri)
+      |> Enum.map(fn {uri, edits} ->
+        %{
+          "textDocument" => %{
+            "uri" => uri,
+            "version" => nil
+          },
+          "edits" =>
+            Enum.map(edits, fn edit ->
+              %{"range" => edit.range, "newText" => new_name}
+            end)
+        }
+      end)
+
+    {:ok, %{"documentChanges" => changes}}
+  end
+
+  def prepare(%SourceFile{} = source_file, _uri, line, character) do
+    result =
+      with %{
+             begin: {start_line, start_col},
+             end: {end_line, end_col},
+             char_ident: char_ident
+           } = res
+           when not is_nil(res) <-
+             get_begin_end_and_char_ident(source_file.text, line, character) do
+        %{
+          range: adjust_range(start_line, start_col, end_line, end_col),
+          placeholder: to_string(char_ident)
+        }
+      else
+        _ ->
+          # Not a variable or local call, skipping for now
+          nil
+      end
+
+    {:ok, result}
+  end
+
+  defp repack_references(references, start_uri) do
+    for reference <- references do
+      uri = if reference.uri, do: SourceFile.Path.to_uri(reference.uri), else: start_uri
+
+      %{
+        uri: uri,
+        range: %{
+          end: %{character: reference.range.end.column - 1, line: reference.range.end.line - 1},
+          start: %{
+            character: reference.range.start.column - 1,
+            line: reference.range.start.line - 1
+          }
+        }
+      }
+    end
+  end
+
+  defp parse_definition_source_code(%{file: file}) do
+    ElixirSense.Core.Parser.parse_file(file, true, true, nil)
+  end
+
+  defp parse_definition_source_code(source_text) when is_binary(source_text) do
+    ElixirSense.Core.Parser.parse_string(source_text, true, true, nil)
+  end
+
+  defp get_all_fn_header_positions(parsed_source, char_ident) do
+    parsed_source.mods_funs_to_positions
+    |> Map.filter(fn
+      {{_, fn_name, _}, _} -> Atom.to_charlist(fn_name) == char_ident
```

Comment:
There's a bug atm - it renames all function headers regardless of arity. It should also match on the arity of the function header

Reply:
Sorted